src/node/hooks/express/webaccess.js

```diff
@@ -70,28 +70,34 @@ const checkAccess = async (req, res, next) => {
   // This helper is used in steps 2 and 4 below, so it may be called twice per access: once before
   // authentication is checked and once after (if settings.requireAuthorization is true).
   const authorize = async () => {
-    const grant = (level) => {
+    const grant = async (level) => {
       level = exports.normalizeAuthzLevel(level);
       if (!level) return false;
       const user = req.session.user;
       if (user == null) return true; // This will happen if authentication is not required.
       const encodedPadId = (req.path.match(/^\/p\/([^/]*)/) || [])[1];
       if (encodedPadId == null) return true;
-      const padId = decodeURIComponent(encodedPadId);
+      let padId = decodeURIComponent(encodedPadId);
+      if (readOnlyManager.isReadOnlyId(padId)) {
+        // pad is read-only, first get the real pad ID
+        padId = await readOnlyManager.getPadId(padId);
```

Comment:
The alternative to performing normalization here would be ensuring that checkAccess() is never called with read-only pad IDs. As far as I can tell, this would currently only affect src/node/padaccess.js (which does not get the real pad ID before querying checkAccess) which is used by pad export.
However, I have decided to perform normalization here in order to prevent similar bugs in the future if developers would forget to "normalize" the pad ID elsewhere.